synapse/storage/databases/main/roommember.py

```diff
@@ -212,9 +212,9 @@ def get_users_in_room_txn(self, txn: LoggingTransaction, room_id: str) -> List[s
         txn.execute(sql, (room_id, Membership.JOIN))
         return [r[0] for r in txn]
 
-    @cached(max_entries=100000, iterable=True)
+    @cached(max_entries=100000, iterable=True, uncached_args=["user_ids"])
```

Comment:
`uncached_args` doesn't quite do the right thing here, as it *excluded* `user_ids` from the key used to cache the function call, so a call to get_users_in_room_with_profiles` for the same room but different sets of users will return the cached value from the first call.

Instead, we probably want to have a separate cache for this call path and use `@cachedList`, e.g. like here:

https://github.com/matrix-org/synapse/blob/d102ad67fddc650c34baa89dc7b2926d46a9aeca/synapse/storage/databases/main/keys.py#L38-L49

This *does* mean that we'd end up with two separate caches, rather than sharing one, but in this case I don't think that's a big deal as other calls to `get_users_in_room_with_profiles` are rare (and the alternative is very fiddly to write).

Reply:
So this I am still a little confused on-everything else should be sorted. I could either hop on a call tomorrow or maybe another explanation might help?